exploration/expression-attributes.md

```diff
@@ -204,38 +217,125 @@ the reserved/private-use rules will need to be adjusted to support attributes.
+option-like `@key=value` attribute pairs with a literal value
+at the end of any expression or markup.
+
+To distinguish attributes from options,
 require `@` as a prefix for each attribute asignment.
-Examples: `@translate=yes` and `@locale=$exprLocale`.
+Examples: `@translate=yes` and `@xliff:canCopy`.
 
-Define the meaning and supported values of some expression attributes in the specification,
-including at least `@dir` and `@locale`.
+Do not allow expression or markup attributes to influence the formatting context,
+or pass them to function handlers.
 
-To support later extension of the specified set of attributes while allowing user extensibility,
-suggest custom attribute names to include a U+002D Hyphen-Minus `-`.
-Examples: `@can-copy=no`, `@note-link=|https://...|`.
+Drop variable values from the `attribute` rule:
 
-Allow expression attributes to influence the formatting context,
-but do not directly pass them to user-defined functions.
+```diff
+-attribute = "@" identifier [[s] "=" [s] (literal / variable)]
++attribute = "@" identifier [[s] "=" [s] literal]
+```
 
 ## Alternatives Considered
 
-### Do not support expression attributes
+### Do nothing
+
+Continue to [caution](https://github.com/unicode-org/message-format-wg/blob/d38ff326d2381b3ef361e996c3431d1b251518d6/spec/syntax.md#attributes)
+function authors and other implementers away from creating function-specific or implementation-specific option values
+for the use cases presented above.
+
+As should be obvious, the current situation is not tenable in the long term, and should be resolved.
+
+### Do not provide any guidance
+
+Do not include in the spec rules or guidance for declaring formatted part identifiers,
+or overriding the message locale or directionality.
+
+Do not define a common way to communicate information
+about an expression or markup to translators or tools.
+
+This would mean not defining anything for default registry functions either,
+effectively requiring implementation-specific options like `icu:locale`.
+
+Other functions could use their own definitions and handling for similar options,
+such as `locale` or `x:lang`.
+
+Formatted parts for markup would not be able to directly include an identifier.
 
 If not explicitly defined, less information will be provided to translators.
 
 Function options may be used as a workaround,
 but each implementation and user will end up with different practices.
 
-### Use function options, but with some suggested prefix like `_`
+### Define options for default registry only
 
-A bit less bad than the previous, but still mixes attributes and options into the same namespace.
+Define at least `locale` and `dir` as options for default registry functions,
+with handling internal to each function implementation.
+
+Other functions could use their own definitions and handling for similar options,
+such as `locale` or `x:lang`.
+
+Formatted parts for markup would not be able to directly include an identifier.
```

Comment:
I don't follow this statement. Why wouldn't I be able to have a `{#b id=foo}`? Or a `{$x :number id=foo}`?

Reply:
This statement is specifically about formatted parts for markup, not expressions. Functions can define how expressions get formatted, but markup is not configurable. Sure, you could pass markup an `id=foo` option, but that option would be only present in the general bag of options in the formatted part, and would potentially conflict with other use.

For example, it's rather common for HTML elements to include an `id` value, which is already relevant for the CSS or JS executing on that document, but an identifier may be separately relevant for the post-processing of a formatted message, such as adding in parameters that were not included in the translation.

In other words, if we're translating a message like
```
Would you like to {#a u:id=continue}accept{/a} or {#a u:id=cancel}reject{/a} the offer?
```
we need a way to disambiguate the two formatted `{#a}` spans from each other, hence the `u:id`. That way the message can end up first formatted as:
```js
[
  { type: 'literal', value: 'Would you like to ' },
  { type: 'markup', kind: 'open', source: '#a', name: 'a', id: 'continue' },
  { type: 'literal', value: 'accept' },
  { type: 'markup', kind: 'close', source: '/a', name: 'a' },
  { type: 'literal', value: ' or ' },
  { type: 'markup', kind: 'open', source: '#a', name: 'a', 'id': 'cancel' },
  { type: 'literal', value: 'reject' },
  { type: 'markup', kind: 'close', source: '/a', name: 'a' },
  { type: 'literal', value: ' the offer?' }
]
```

And from there, we can construct an HTML snippet with additional properties:
```html
Would you like to <a href="continue?token=123abc">accept</a> or <a href="cancel?token=123abc">reject</a> the offer?
```

But if there's no common `u:id` definition, then the identifier ends up mixed together with the other markup options, and it gets much trickier to build tooling that's universal across different implementations and function authors, and doesn't risk bleeding through into the output HTML, where a property like `id="continue"` could very easily conflict with some other operations and expectations of the page.

Reply:
Okay, I see what you're saying now. My example omitted the `u:` by mistake. I still think that the text you have the design document is super-mysterious, though. Note that attributed strings can have overlapping markup (that would be illegal in HTML). For example:

> Would you like to {#span type=foo u:id=1}click {#span type=bar u:id=2}here{/span u:id=1} to get {/span u:id=2} illegal stuff.


Perhaps:

```suggestion
Lack of a built-in identifier would inhibit tools support for markup,
particularly when formatted to parts, 
when the relationship of open and close markup is ambiguous.
> For example, the two `{/a}` markup elements in this mess:
>```
> Would you like to {#a u:id=continue}accept{/a} or {#a u:id=cancel}reject{/a} the offer?
>```
```

Reply:
I think that a reference to "tools support" here would be misleading, as we're talking about the actual formatted parts. Also, as markup formats as an empty string, anything relating to it is only relevant with formatted parts or similar non-concatenated output formats.

In my experience, needing to rely on something like `u:id` to detangle improper nesting is relatively rare compared to needing to allow for localizations to reorder actions such as "accept" and "cancel"; in my previous comment's example I included a `u:id` to show how that can be supported.

```suggestion
Formatted parts for markup would not be able to directly include an identifier.
Implementations and users will need to invent their own practices with
markup option names like `l10n-id`, `mf:id`, or `markupId`
to refer to specific markup parts.
```